File: skills/clean-ddd-hexagonal/examples/go/domain/entity.go
// Package shared provides base building blocks for domain-driven design.
// It contains Entity and AggregateRoot types that form the foundation
// for all domain models following DDD tactical patterns.
package shared

// Entity represents a domain object with a unique identity.
// Entities are distinguished by their identity rather than their attributes.
// Two entities with the same ID are considered equal, regardless of their
// other attribute values.
//
// The type parameter ID must be comparable to support identity comparison.
type Entity[ID comparable] struct {
	id ID
}

// NewEntity creates a new Entity with the specified identifier.
// This constructor should be used when creating new domain entities
// that require identity tracking.
func NewEntity[ID comparable](id ID) Entity[ID] {
	return Entity[ID]{id: id}
}

// ID returns the unique identifier of this entity.
func (e Entity[ID]) ID() ID {
	return e.id
}

// Equals compares two entities by their identity.
// Returns true if both entities have the same ID, regardless of
// their other attribute values.
func (e Entity[ID]) Equals(other Entity[ID]) bool {
	return e.id == other.id
}

File: skills/clean-ddd-hexagonal/examples/typescript/domain/entity.ts
// domain/shared/entity.ts

/**
 * Base class for domain objects that have a distinct identity.
 *
 * Entities are distinguished by their identity (ID) rather than their attributes.
 * Two entities are considered equal if they have the same identity, regardless
 * of their other attribute values.
 *
 * @typeParam T - The type of the entity's identifier
 *
 * @example
 * ```typescript
 * class User extends Entity<UserId> {
 *   constructor(id: UserId, private name: string) {
 *     super(id);
 *   }
 * }
 * ```
 */
export abstract class Entity<T> {
  /** The unique identifier for this entity. */
  protected readonly _id: T;

  /**
   * Create a new entity with the specified identifier.
   * @param id - The unique identifier for this entity
   */
  protected constructor(id: T) {
    this._id = id;
  }

  /**
   * Get the unique identifier of this entity.
   * @returns The entity's identifier
   */
  get id(): T {
    return this._id;
  }

  /**
   * Compare this entity with another by identity.
   *
   * @param other - The entity to compare against
   * @returns True if both entities have the same identity
   */
  equals(other: Entity<T>): boolean {
    if (other === null || other === undefined) {
      return false;
    }
    if (!(other instanceof Entity)) {
      return false;
    }
    return this._id === other._id;
  }
}
